File: Server/src/services/tools/manage_transform.py
"""
Transform management tool for Unity GameObjects.

Provides operations for getting/setting world and local transforms,
bounds queries, alignment, distribution, and placement validation.
Part of the 'spatial' tool group for advanced scene construction.
"""
from typing import Annotated, Any, Literal

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from services.tools.action_policy import maybe_run_tool_preflight
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry
from services.tools.utils import (
    coerce_bool,
    coerce_float,
    normalize_vector3,
    normalize_string_list,
)


@mcp_for_unity_tool(
    name="manage_transform",
    description=(
        "Advanced transform operations for GameObjects including world/local space queries, "
        "bounds retrieval, grid snapping, alignment, distribution, and placement validation. "
        "Complements manage_gameobject by focusing purely on spatial manipulation."
    ),
    annotations=ToolAnnotations(
        title="Manage Transform",
        destructiveHint=True,
    ),
    group="spatial",
)
async def manage_transform(
    ctx: Context,
    action: Annotated[
        Literal[
            "get_world_transform",
            "get_local_transform",
            "set_world_transform",
            "set_local_transform",
            "get_bounds",
            "snap_to_grid",
            "align_to_object",
            "distribute_objects",
            "place_relative",
            "validate_placement",
        ],
        "Transform action to perform",
    ],
    target: Annotated[
        str | None,
        "Target GameObject identifier (name, path, instance ID, or null for selected)"
    ] = None,
    search_method: Annotated[
        Literal["by_id", "by_name", "by_path", "by_tag", "by_layer", "by_component"],
        "How to resolve 'target' if provided as string"
    ] = "by_name",
    # Transform data parameters
    position: Annotated[
        list[float] | dict[str, float] | str | None,
        "Position as [x, y, z] array, {x, y, z} object, or JSON string"
    ] = None,
    rotation: Annotated[
        list[float] | dict[str, float] | str | None,
        "Rotation as [x, y, z] euler angles array, {x, y, z} object, or JSON string"
    ] = None,
    scale: Annotated[
        list[float] | dict[str, float] | str | None,
        "Scale as [x, y, z] array, {x, y, z} object, or JSON string"
    ] = None,
    # Grid snapping parameters
    grid_size: Annotated[
        float | str | None,
        "Grid size for snap_to_grid (default: 1.0)"
    ] = None,
    snap_position: Annotated[
        bool | str | None,
        "Whether to snap position (default: True)"
    ] = None,
    snap_rotation: Annotated[
        bool | str | None,
        "Whether to snap rotation to 90-degree increments (default: False)"
    ] = None,
    # Alignment parameters
    reference_object: Annotated[
        str | None,
        "Reference GameObject for alignment operations"
    ] = None,
    align_axis: Annotated[
        Literal["x", "y", "z", "all"],
        "Axis to align (default: all)"
    ] = "all",
    align_mode: Annotated[
        Literal["min", "max", "center", "pivot"],
        "Alignment mode: align to min bounds, max bounds, center, or pivot (default: center)"
    ] = "center",
    # Distribution parameters
    targets: Annotated[
        list[str] | str | None,
        "List of target GameObject identifiers for distribute_objects"
    ] = None,
    distribute_axis: Annotated[
        Literal["x", "y", "z"],
        "Axis for distribution (default: x)"
    ] = "x",
    distribute_spacing: Annotated[
        float | str | None,
        "Spacing between objects for distribution (default: auto-calculate)"
    ] = None,
    # Relative placement parameters
    offset: Annotated[
        list[float] | dict[str, float] | str | None,
        "Offset vector for place_relative as [x, y, z]"
    ] = None,
    direction: Annotated[
        Literal[
            "left", "right", "up", "down", "forward", "back",
            "front", "backward", "behind", "above", "below",
            "north", "south", "east", "west"
        ],
        "Cardinal direction for relative placement"
    ] = None,
    distance: Annotated[
        float | str | None,
        "Distance in the specified direction (default: 1.0)"
    ] = None,
    use_world_space: Annotated[
        bool | str | None,
        "Use world space directions (default: True) or reference object's local space"
    ] = None,
    # Validation parameters
    check_overlap: Annotated[
        bool | str | None,
        "Check for overlapping objects in validate_placement (default: True)"
    ] = None,
    check_off_grid: Annotated[
        bool | str | None,
        "Check if position is off-grid in validate_placement (default: True)"
    ] = None,
    check_invalid_scale: Annotated[
        bool | str | None,
        "Check for invalid (zero/negative) scale in validate_placement (default: True)"
    ] = None,
    min_spacing: Annotated[
        float | str | None,
        "Minimum spacing required between objects for overlap check (default: 0.0)"
    ] = None,
) -> dict[str, Any]:
    """
    Perform advanced transform operations on GameObjects.

    Actions:
    - get_world_transform: Get world-space position, rotation, and scale
    - get_local_transform: Get local-space position, rotation, and scale
    - set_world_transform: Set world-space transform (position, rotation, scale)
    - set_local_transform: Set local-space transform
    - get_bounds: Get renderer/collider bounds (center, extents, min, max)
    - snap_to_grid: Snap object position/rotation to grid
    - align_to_object: Align to another object by bounds or pivot
    - distribute_objects: Evenly distribute multiple objects along an axis
    - place_relative: Place object relative to another with offset/direction
    - validate_placement: Check for off-grid, overlap, and invalid scale
    """
    unity_instance = await get_unity_instance_from_context(ctx)

    gate = await maybe_run_tool_preflight(ctx, "manage_transform")
    if gate is not None:
        return gate.model_dump()

    # Validate required parameters
    if action is None:
        return {
            "success": False,
            "message": "Missing required parameter 'action'. Valid actions: get_world_transform, get_local_transform, "
                      "set_world_transform, set_local_transform, get_bounds, snap_to_grid, align_to_object, "
                      "distribute_objects, place_relative, validate_placement"
        }

    # Normalize vector parameters
    position, pos_error = normalize_vector3(position, "position")
    if pos_error:
        return {"success": False, "message": pos_error}

    rotation, rot_error = normalize_vector3(rotation, "rotation")
    if rot_error:
        return {"success": False, "message": rot_error}

    scale, scale_error = normalize_vector3(scale, "scale")
    if scale_error:
        return {"success": False, "message": scale_error}

    offset, offset_error = normalize_vector3(offset, "offset")
    if offset_error:
        return {"success": False, "message": offset_error}

    # Normalize boolean parameters
    snap_position = coerce_bool(snap_position, default=True)
    snap_rotation = coerce_bool(snap_rotation, default=False)
    use_world_space = coerce_bool(use_world_space, default=True)
    check_overlap = coerce_bool(check_overlap, default=True)
    check_off_grid = coerce_bool(check_off_grid, default=True)
    check_invalid_scale = coerce_bool(check_invalid_scale, default=True)

    # Normalize float parameters
    grid_size = coerce_float(grid_size, default=1.0)
    distribute_spacing = coerce_float(distribute_spacing)
    distance = coerce_float(distance, default=1.0)
    min_spacing = coerce_float(min_spacing, default=0.0)

    # Normalize targets list for distribute_objects
    targets_list, targets_error = normalize_string_list(targets, "targets")
    if targets_error:
        return {"success": False, "message": targets_error}

    try:
        params = {
            "action": action,
            "target": target,
            "searchMethod": search_method,
            "position": position,
            "rotation": rotation,
            "scale": scale,
            "gridSize": grid_size,
            "snapPosition": snap_position,
            "snapRotation": snap_rotation,
            "referenceObject": reference_object,
            "alignAxis": align_axis,
            "alignMode": align_mode,
            "targets": targets_list,
            "distributeAxis": distribute_axis,
            "distributeSpacing": distribute_spacing,
            "offset": offset,
            "direction": direction,
            "distance": distance,
            "useWorldSpace": use_world_space,
            "checkOverlap": check_overlap,
            "checkOffGrid": check_off_grid,
            "checkInvalidScale": check_invalid_scale,
            "minSpacing": min_spacing,
        }
        
        # Remove None values
        params = {k: v for k, v in params.items() if v is not None}

        response = await send_with_unity_instance(
            async_send_command_with_retry,
            unity_instance,
            "manage_transform",
            params,
        )

        if isinstance(response, dict) and response.get("success"):
            return {
                "success": True,
                "message": response.get("message", f"Transform operation '{action}' completed successfully."),
                "data": response.get("data")
            }
        return response if isinstance(response, dict) else {"success": False, "message": str(response)}

    except Exception as e:
        return {"success": False, "message": f"Python error in manage_transform: {e!s}"}

File: MCPForUnity/Editor/Tools/Spatial/ManageTransform.cs
#nullable disable
using System;
using System.Collections.Generic;
using System.Linq;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace MCPForUnity.Editor.Tools.Spatial
{
    /// <summary>
    /// Handles advanced transform operations for GameObjects including world/local space queries,
    /// bounds retrieval, grid snapping, alignment, distribution, and placement validation.
    /// Part of the 'spatial' tool group.
    /// </summary>
    [McpForUnityTool("manage_transform", AutoRegister = false, Group = "spatial")]
    public static class ManageTransform
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
            {
                return new ErrorResponse("Parameters cannot be null.");
            }

            string action = @params["action"]?.ToString().ToLower();
            if (string.IsNullOrEmpty(action))
            {
                return new ErrorResponse("Action parameter is required.");
            }

            // Get target GameObject (or use selected if not specified)
            JToken targetToken = @params["target"];
            string searchMethod = @params["searchMethod"]?.ToString().ToLower();

            GameObject targetGo = null;
            if (targetToken != null)
            {
                targetGo = FindGameObject(targetToken, searchMethod);
                if (targetGo == null)
                {
                    return new ErrorResponse($"Target GameObject '{targetToken}' not found.");
                }
            }
            else
            {
                // Use current selection if available
                if (Selection.activeGameObject != null)
                {
                    targetGo = Selection.activeGameObject;
                }
                else
                {
                    return new ErrorResponse("No target specified and no GameObject selected.");
                }
            }

            try
            {
                switch (action)
                {
                    case "get_world_transform":
                        return GetWorldTransform(targetGo);
                    case "get_local_transform":
                        return GetLocalTransform(targetGo);
                    case "set_world_transform":
                        return SetWorldTransform(targetGo, @params);
                    case "set_local_transform":
                        return SetLocalTransform(targetGo, @params);
                    case "get_bounds":
                        return GetBounds(targetGo);
                    case "snap_to_grid":
                        return SnapToGrid(targetGo, @params);
                    case "align_to_object":
                        return AlignToObject(targetGo, @params);
                    case "distribute_objects":
                        return DistributeObjects(@params);
                    case "place_relative":
                        return PlaceRelative(targetGo, @params);
                    case "validate_placement":
                        return ValidatePlacement(targetGo, @params);

                    default:
                        return new ErrorResponse($"Unknown action: '{action}'. Valid actions: get_world_transform, get_local_transform, set_world_transform, set_local_transform, get_bounds, snap_to_grid, align_to_object, distribute_objects, place_relative, validate_placement");
                }
            }
            catch (Exception e)
            {
                McpLog.Error($"[ManageTransform] Action '{action}' failed: {e}");
                return new ErrorResponse($"Internal error processing action '{action}': {e.Message}");
            }
        }

        private static GameObject FindGameObject(JToken targetToken, string searchMethod)
        {
            if (targetToken.Type == JTokenType.Integer)
            {
                int instanceId = targetToken.ToObject<int>();
                return GameObjectLookup.FindById(instanceId);
            }

            string targetStr = targetToken.ToString();
            if (targetStr == "selected")
            {
                return Selection.activeGameObject;
            }

            var results = GameObjectLookup.SearchGameObjects(searchMethod, targetStr, true, 1);
            return results.Count > 0 ? GameObjectLookup.FindById(results[0]) : null;
        }

        private static object GetWorldTransform(GameObject go)
        {
            var t = go.transform;
            var data = new
            {
                instanceId = go.GetInstanceID(),
                name = go.name,
                position = new[] { t.position.x, t.position.y, t.position.z },
                rotation = new[] { t.eulerAngles.x, t.eulerAngles.y, t.eulerAngles.z },
                scale = new[] { t.lossyScale.x, t.lossyScale.y, t.lossyScale.z },
                space = "world"
            };

            return new SuccessResponse($"World transform for '{go.name}' retrieved.", data);
        }

        private static object GetLocalTransform(GameObject go)
        {
            var t = go.transform;
            var data = new
            {
                instanceId = go.GetInstanceID(),
                name = go.name,
                position = new[] { t.localPosition.x, t.localPosition.y, t.localPosition.z },
                rotation = new[] { t.localEulerAngles.x, t.localEulerAngles.y, t.localEulerAngles.z },
                scale = new[] { t.localScale.x, t.localScale.y, t.localScale.z },
                parent = t.parent?.name,
                space = "local"
            };

            return new SuccessResponse($"Local transform for '{go.name}' retrieved.", data);
        }

        private static object SetWorldTransform(GameObject go, JObject @params)
        {
            Vector3? position = VectorParsing.ParseVector3(@params["position"]);
            Vector3? rotation = VectorParsing.ParseVector3(@params["rotation"]);
            Vector3? scale = VectorParsing.ParseVector3(@params["scale"]);

            Undo.RecordObject(go.transform, "Set World Transform");

            var t = go.transform;
            if (position.HasValue) t.position = position.Value;
            if (rotation.HasValue) t.rotation = Quaternion.Euler(rotation.Value);
            if (scale.HasValue)
            {
                // World scale is read-only in Unity, so we warn if user tries to set it
                return new ErrorResponse("Cannot set world scale directly. Use set_local_transform with scale parameter, or modify the object's transform hierarchy.");
            }

            EditorUtility.SetDirty(go);

            var data = new
            {
                position = new[] { t.position.x, t.position.y, t.position.z },
                rotation = new[] { t.eulerAngles.x, t.eulerAngles.y, t.eulerAngles.z }
            };

            return new SuccessResponse($"World transform updated for '{go.name}'.", data);
        }

        private static object SetLocalTransform(GameObject go, JObject @params)
        {
            Vector3? position = VectorParsing.ParseVector3(@params["position"]);
            Vector3? rotation = VectorParsing.ParseVector3(@params["rotation"]);
            Vector3? scale = VectorParsing.ParseVector3(@params["scale"]);

            Undo.RecordObject(go.transform, "Set Local Transform");

            var t = go.transform;
            if (position.HasValue) t.localPosition = position.Value;
            if (rotation.HasValue) t.localEulerAngles = rotation.Value;
            if (scale.HasValue) t.localScale = scale.Value;

            EditorUtility.SetDirty(go);

            var data = new
            {
                position = new[] { t.localPosition.x, t.localPosition.y, t.localPosition.z },
                rotation = new[] { t.localEulerAngles.x, t.localEulerAngles.y, t.localEulerAngles.z },
                scale = new[] { t.localScale.x, t.localScale.y, t.localScale.z }
            };

            return new SuccessResponse($"Local transform updated for '{go.name}'.", data);
        }

        private static object GetBounds(GameObject go)
        {
            var renderers = go.GetComponentsInChildren<Renderer>(true);
            var colliders = go.GetComponentsInChildren<Collider>(true);

            if (renderers.Length == 0 && colliders.Length == 0)
            {
                // Return bounds based on transform only
                var data = new
                {
                    hasRenderer = false,
                    hasCollider = false,
                    center = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z },
                    extents = new[] { 0f, 0f, 0f },
                    size = new[] { 0f, 0f, 0f },
                    min = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z },
                    max = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z }
                };
                return new SuccessResponse($"No renderer or collider found for '{go.name}'. Returning pivot position.", data);
            }

            Bounds bounds = new Bounds(go.transform.position, Vector3.zero);
            bool hasBounds = false;

            foreach (var renderer in renderers)
            {
                if (!hasBounds)
                {
                    bounds = renderer.bounds;
                    hasBounds = true;
                }
                else
                {
                    bounds.Encapsulate(renderer.bounds);
                }
            }

            foreach (var collider in colliders)
            {
                if (!hasBounds)
                {
                    bounds = collider.bounds;
                    hasBounds = true;
                }
                else
                {
                    bounds.Encapsulate(collider.bounds);
                }
            }

            var boundsData = new
            {
                hasRenderer = renderers.Length > 0,
                rendererCount = renderers.Length,
                hasCollider = colliders.Length > 0,
                colliderCount = colliders.Length,
                center = new[] { bounds.center.x, bounds.center.y, bounds.center.z },
                extents = new[] { bounds.extents.x, bounds.extents.y, bounds.extents.z },
                size = new[] { bounds.size.x, bounds.size.y, bounds.size.z },
                min = new[] { bounds.min.x, bounds.min.y, bounds.min.z },
                max = new[] { bounds.max.x, bounds.max.y, bounds.max.z }
            };

            return new SuccessResponse($"Bounds retrieved for '{go.name}'.", boundsData);
        }

        private static object SnapToGrid(GameObject go, JObject @params)
        {
            float gridSize = @params["gridSize"]?.ToObject<float>() ?? 1.0f;
            bool snapPosition = @params["snapPosition"]?.ToObject<bool>() ?? true;
            bool snapRotation = @params["snapRotation"]?.ToObject<bool>() ?? false;

            Undo.RecordObject(go.transform, "Snap to Grid");

            var t = go.transform;

            if (snapPosition)
            {
                Vector3 snappedPos = new Vector3(
                    Mathf.Round(t.position.x / gridSize) * gridSize,
                    Mathf.Round(t.position.y / gridSize) * gridSize,
                    Mathf.Round(t.position.z / gridSize) * gridSize
                );
                t.position = snappedPos;
            }

            if (snapRotation)
            {
                Vector3 snappedRot = new Vector3(
                    Mathf.Round(t.eulerAngles.x / 90f) * 90f,
                    Mathf.Round(t.eulerAngles.y / 90f) * 90f,
                    Mathf.Round(t.eulerAngles.z / 90f) * 90f
                );
                t.rotation = Quaternion.Euler(snappedRot);
            }

            EditorUtility.SetDirty(go);

            var data = new
            {
                gridSize,
                snappedPosition = new[] { t.position.x, t.position.y, t.position.z },
                snappedRotation = new[] { t.eulerAngles.x, t.eulerAngles.y, t.eulerAngles.z }
            };

            return new SuccessResponse($"'{go.name}' snapped to grid (size: {gridSize}).", data);
        }

        private static object AlignToObject(GameObject go, JObject @params)
        {
            string referenceId = @params["referenceObject"]?.ToString();
            if (string.IsNullOrEmpty(referenceId))
            {
                return new ErrorResponse("referenceObject parameter is required for align_to_object.");
            }

            string searchMethod = @params["searchMethod"]?.ToString().ToLower();
            GameObject referenceGo = FindGameObject(referenceId, searchMethod);
            if (referenceGo == null)
            {
                return new ErrorResponse($"Reference object '{referenceId}' not found.");
            }

            string alignAxis = @params["alignAxis"]?.ToString().ToLower() ?? "all";
            string alignMode = @params["alignMode"]?.ToString().ToLower() ?? "center";

            Undo.RecordObject(go.transform, "Align to Object");

            Vector3 targetPos = CalculateAlignPosition(go, referenceGo, alignAxis, alignMode);
            go.transform.position = targetPos;

            EditorUtility.SetDirty(go);

            var data = new
            {
                alignedTo = referenceGo.name,
                alignAxis,
                alignMode,
                newPosition = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z }
            };

            return new SuccessResponse($"'{go.name}' aligned to '{referenceGo.name}' ({alignAxis} axis, {alignMode} mode).", data);
        }

        private static Vector3 CalculateAlignPosition(GameObject source, GameObject target, string axis, string mode)
        {
            Bounds sourceBounds = CalculateBounds(source);
            Bounds targetBounds = CalculateBounds(target);

            Vector3 newPos = source.transform.position;

            if (mode == "pivot")
            {
                // Align pivots
                if (axis == "all" || axis == "x") newPos.x = target.transform.position.x;
                if (axis == "all" || axis == "y") newPos.y = target.transform.position.y;
                if (axis == "all" || axis == "z") newPos.z = target.transform.position.z;
            }
            else if (mode == "center")
            {
                // Align centers
                if (axis == "all" || axis == "x") newPos.x = targetBounds.center.x + (newPos.x - sourceBounds.center.x);
                if (axis == "all" || axis == "y") newPos.y = targetBounds.center.y + (newPos.y - sourceBounds.center.y);
                if (axis == "all" || axis == "z") newPos.z = targetBounds.center.z + (newPos.z - sourceBounds.center.z);
            }
            else if (mode == "min")
            {
                // Align min bounds
                if (axis == "all" || axis == "x") newPos.x = targetBounds.min.x + (newPos.x - sourceBounds.min.x);
                if (axis == "all" || axis == "y") newPos.y = targetBounds.min.y + (newPos.y - sourceBounds.min.y);
                if (axis == "all" || axis == "z") newPos.z = targetBounds.min.z + (newPos.z - sourceBounds.min.z);
            }
            else if (mode == "max")
            {
                // Align max bounds
                if (axis == "all" || axis == "x") newPos.x = targetBounds.max.x - (sourceBounds.max.x - newPos.x);
                if (axis == "all" || axis == "y") newPos.y = targetBounds.max.y - (sourceBounds.max.y - newPos.y);
                if (axis == "all" || axis == "z") newPos.z = targetBounds.max.z - (sourceBounds.max.z - newPos.z);
            }

            return newPos;
        }

        private static Bounds CalculateBounds(GameObject go)
        {
            var renderers = go.GetComponentsInChildren<Renderer>(true);
            var colliders = go.GetComponentsInChildren<Collider>(true);

            Bounds bounds = new Bounds(go.transform.position, Vector3.zero);
            bool hasBounds = false;

            foreach (var renderer in renderers)
            {
                if (!hasBounds)
                {
                    bounds = renderer.bounds;
                    hasBounds = true;
                }
                else
                {
                    bounds.Encapsulate(renderer.bounds);
                }
            }

            foreach (var collider in colliders)
            {
                if (!hasBounds)
                {
                    bounds = collider.bounds;
                    hasBounds = true;
                }
                else
                {
                    bounds.Encapsulate(collider.bounds);
                }
            }

            // If no bounds found, use a point at the transform position
            if (!hasBounds)
            {
                bounds = new Bounds(go.transform.position, Vector3.zero);
            }

            return bounds;
        }

        private static object DistributeObjects(JObject @params)
        {
            JArray targetsArray = @params["targets"] as JArray;
            if (targetsArray == null || targetsArray.Count < 2)
            {
                return new ErrorResponse("At least 2 target objects are required for distribute_objects.");
            }

            string searchMethod = @params["searchMethod"]?.ToString().ToLower();
            string distributeAxis = @params["distributeAxis"]?.ToString().ToLower() ?? "x";
            float? distributeSpacing = @params["distributeSpacing"]?.ToObject<float?>();

            // Resolve all target objects
            List<GameObject> objects = new List<GameObject>();
            foreach (var token in targetsArray)
            {
                GameObject go = FindGameObject(token, searchMethod);
                if (go == null)
                {
                    return new ErrorResponse($"Target object '{token}' not found.");
                }
                objects.Add(go);
            }

            // Sort objects by position along the distribution axis
            int axisIndex = distributeAxis switch
            {
                "x" => 0,
                "y" => 1,
                "z" => 2,
                _ => 0
            };

            objects.Sort((a, b) =>
            {
                float posA = axisIndex == 0 ? a.transform.position.x : axisIndex == 1 ? a.transform.position.y : a.transform.position.z;
                float posB = axisIndex == 0 ? b.transform.position.x : axisIndex == 1 ? b.transform.position.y : b.transform.position.z;
                return posA.CompareTo(posB);
            });

            // Calculate positions
            float startPos, endPos, spacing;
            if (distributeSpacing.HasValue)
            {
                spacing = distributeSpacing.Value;
                startPos = axisIndex == 0 ? objects[0].transform.position.x :
                          axisIndex == 1 ? objects[0].transform.position.y : objects[0].transform.position.z;
            }
            else
            {
                // Auto-calculate spacing based on first and last object
                startPos = axisIndex == 0 ? objects[0].transform.position.x :
                          axisIndex == 1 ? objects[0].transform.position.y : objects[0].transform.position.z;
                endPos = axisIndex == 0 ? objects[objects.Count - 1].transform.position.x :
                        axisIndex == 1 ? objects[objects.Count - 1].transform.position.y : objects[objects.Count - 1].transform.position.z;
                spacing = (endPos - startPos) / (objects.Count - 1);
            }

            // Record undos and apply positions
            Undo.SetCurrentGroupName("Distribute Objects");
            int undoGroup = Undo.GetCurrentGroup();

            for (int i = 0; i < objects.Count; i++)
            {
                Undo.RecordObject(objects[i].transform, $"Distribute {objects[i].name}");
                Vector3 pos = objects[i].transform.position;
                float newPos = startPos + (i * spacing);

                if (axisIndex == 0) pos.x = newPos;
                else if (axisIndex == 1) pos.y = newPos;
                else pos.z = newPos;

                objects[i].transform.position = pos;
                EditorUtility.SetDirty(objects[i]);
            }

            Undo.CollapseUndoOperations(undoGroup);

            var data = new
            {
                objectCount = objects.Count,
                distributeAxis,
                spacing,
                objects = objects.Select(go => new { instanceId = go.GetInstanceID(), name = go.name }).ToArray()
            };

            return new SuccessResponse($"Distributed {objects.Count} objects along {distributeAxis} axis (spacing: {spacing:F3}).", data);
        }

        private static object PlaceRelative(GameObject go, JObject @params)
        {
            string referenceId = @params["referenceObject"]?.ToString();
            if (string.IsNullOrEmpty(referenceId))
            {
                return new ErrorResponse("referenceObject parameter is required for place_relative.");
            }

            string searchMethod = @params["searchMethod"]?.ToString().ToLower();
            GameObject referenceGo = FindGameObject(referenceId, searchMethod);
            if (referenceGo == null)
            {
                return new ErrorResponse($"Reference object '{referenceId}' not found.");
            }

            Vector3? offset = VectorParsing.ParseVector3(@params["offset"]);
            string direction = @params["direction"]?.ToString().ToLower();
            float distance = @params["distance"]?.ToObject<float>() ?? 1.0f;
            bool useWorldSpace = @params["useWorldSpace"]?.ToObject<bool>() ?? true;

            Undo.RecordObject(go.transform, "Place Relative");

            Vector3 newPosition;
            Transform referenceTransform = referenceGo.transform;

            if (offset.HasValue)
            {
                // Use explicit offset
                if (useWorldSpace)
                {
                    newPosition = referenceTransform.position + offset.Value;
                }
                else
                {
                    newPosition = referenceTransform.TransformPoint(offset.Value);
                }
            }
            else if (!string.IsNullOrEmpty(direction))
            {
                // Use direction and distance
                Vector3 dir = ParseDirection(direction, useWorldSpace ? null : referenceTransform);
                newPosition = referenceTransform.position + dir * distance;
            }
            else
            {
                return new ErrorResponse("Either 'offset' or 'direction' parameter is required for place_relative.");
            }

            go.transform.position = newPosition;
            EditorUtility.SetDirty(go);

            var data = new
            {
                referenceObject = referenceGo.name,
                newPosition = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z },
                relativeTo = referenceTransform.position,
                useWorldSpace
            };

            return new SuccessResponse($"'{go.name}' placed relative to '{referenceGo.name}'.", data);
        }

        private static Vector3 ParseDirection(string direction, Transform localSpace)
        {
            Vector3 worldDir = direction switch
            {
                "left" => Vector3.left,
                "right" => Vector3.right,
                "up" => Vector3.up,
                "down" => Vector3.down,
                "forward" or "front" => Vector3.forward,
                "back" or "backward" or "behind" => Vector3.back,
                "above" => Vector3.up,
                "below" => Vector3.down,
                "north" => Vector3.forward,
                "south" => Vector3.back,
                "east" => Vector3.right,
                "west" => Vector3.left,
                _ => Vector3.forward
            };

            if (localSpace != null)
            {
                return localSpace.TransformDirection(worldDir);
            }

            return worldDir;
        }

        private static object ValidatePlacement(GameObject go, JObject @params)
        {
            bool checkOverlap = @params["checkOverlap"]?.ToObject<bool>() ?? true;
            bool checkOffGrid = @params["checkOffGrid"]?.ToObject<bool>() ?? true;
            bool checkInvalidScale = @params["checkInvalidScale"]?.ToObject<bool>() ?? true;
            float minSpacing = @params["minSpacing"]?.ToObject<float>() ?? 0f;

            var issues = new List<string>();
            var warnings = new List<string>();

            // Check off-grid
            if (checkOffGrid)
            {
                float gridSize = 1.0f; // Default grid size
                Vector3 pos = go.transform.position;
                bool onGridX = Mathf.Abs(pos.x % gridSize) < 0.001f || Mathf.Abs(pos.x % gridSize - gridSize) < 0.001f;
                bool onGridY = Mathf.Abs(pos.y % gridSize) < 0.001f || Mathf.Abs(pos.y % gridSize - gridSize) < 0.001f;
                bool onGridZ = Mathf.Abs(pos.z % gridSize) < 0.001f || Mathf.Abs(pos.z % gridSize - gridSize) < 0.001f;

                if (!onGridX || !onGridY || !onGridZ)
                {
                    warnings.Add($"Object is not aligned to {gridSize} unit grid");
                }
            }

            // Check invalid scale
            if (checkInvalidScale)
            {
                Vector3 scale = go.transform.localScale;
                if (scale.x <= 0 || scale.y <= 0 || scale.z <= 0)
                {
                    issues.Add("Object has zero or negative scale");
                }
            }

            // Check overlap
            if (checkOverlap)
            {
                Bounds myBounds = CalculateBounds(go);
                var allObjects = GameObjectLookup.GetAllSceneObjects(false);

                foreach (var other in allObjects)
                {
                    if (other == go) continue;

                    Bounds otherBounds = CalculateBounds(other);
                    if (myBounds.Intersects(otherBounds))
                    {
                        // Calculate minimum distance between bounds
                        float distance = Vector3.Distance(myBounds.center, otherBounds.center) -
                                        (myBounds.extents.magnitude + otherBounds.extents.magnitude);

                        if (distance < minSpacing)
                        {
                            issues.Add($"Overlaps with '{other.name}' (distance: {distance:F3})");
                        }
                    }
                }
            }

            bool isValid = issues.Count == 0;
            string message = isValid
                ? $"Placement validation passed for '{go.name}'"
                : $"Placement validation found {issues.Count} issue(s) for '{go.name}'";

            if (warnings.Count > 0)
            {
                message += $" and {warnings.Count} warning(s)";
            }

            var data = new
            {
                isValid,
                issues,
                warnings,
                position = new[] { go.transform.position.x, go.transform.position.y, go.transform.position.z },
                scale = new[] { go.transform.localScale.x, go.transform.localScale.y, go.transform.localScale.z }
            };

            if (!isValid)
            {
                return new ErrorResponse(message, data);
            }

            return new SuccessResponse(message, data);
        }
    }
}
